drivers/rtc/rtc_ds1307.c

```diff
@@ -0,0 +1,156 @@
+#define DS1307_REG_HOURS   0x02
+#define DS1307_REG_DAY     0x03
+#define DS1307_REG_DATE    0x04
+#define DS1307_REG_MONTH   0x05
+#define DS1307_REG_YEAR    0x06
+#define DS1307_REG_CTRL    0x07
+
+#define SECONDS_BITS GENMASK(6, 0)
+#define MINUTES_BITS GENMASK(7, 0)
+#define HOURS_BITS   GENMASK(5, 0)
+#define DATE_BITS    GENMASK(5, 0)
+#define MONTHS_BITS  GENMASK(4, 0)
+#define WEEKDAY_BITS GENMASK(2, 0)
+#define YEAR_BITS    GENMASK(7, 0)
+
+struct ds1307_config {
+	struct i2c_dt_spec i2c_bus;
+};
+
+struct ds1307_data {
+	struct k_mutex lock;
+};
+
+static int ds1307_set_time(const struct device *dev, const struct rtc_time *tm)
+{
+	int err;
+	uint8_t regs[7];
+
+	struct ds1307_data *data = dev->data;
+	const struct ds1307_config *config = dev->config;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	LOG_DBG("set time: year = %d, mon = %d, mday = %d, wday = %d, hour = %d, "
+		"min = %d, sec = %d",
+		tm->tm_year, tm->tm_mon, tm->tm_mday, tm->tm_wday, tm->tm_hour, tm->tm_min,
+		tm->tm_sec);
+
+	regs[0] = bin2bcd(tm->tm_sec) & SECONDS_BITS;
+	regs[1] = bin2bcd(tm->tm_min);
+	regs[2] = bin2bcd(tm->tm_hour);
+	regs[3] = bin2bcd(tm->tm_wday);
+	regs[4] = bin2bcd(tm->tm_mday);
+	regs[5] = bin2bcd(tm->tm_mon);
+	regs[6] = bin2bcd(tm->tm_year);
+
+	err = i2c_burst_write_dt(&config->i2c_bus, DS1307_REG_SECONDS, regs, sizeof(regs));
+
+	k_mutex_unlock(&data->lock);
+
+	return err;
+}
+
+static int ds1307_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	int err;
+	uint8_t regs[7];
+
+	struct ds1307_data *data = dev->data;
+	const struct ds1307_config *config = dev->config;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	err = i2c_burst_read_dt(&config->i2c_bus, DS1307_REG_SECONDS, regs, sizeof(regs));
+	if (err != 0) {
+		goto unlock;
+	}
+
+	timeptr->tm_sec = bcd2bin(regs[0] & SECONDS_BITS);
+	timeptr->tm_min = bcd2bin(regs[1] & MINUTES_BITS);
+	timeptr->tm_hour = bcd2bin(regs[2] & HOURS_BITS); /* 24hr mode */
+	timeptr->tm_wday = bcd2bin(regs[3] & WEEKDAY_BITS);
+	timeptr->tm_mday = bcd2bin(regs[4] & DATE_BITS);
+	timeptr->tm_mon = bcd2bin(regs[5] & MONTHS_BITS);
+	timeptr->tm_year = bcd2bin(regs[6] & YEAR_BITS);
+
+	LOG_DBG("get time: year = %d, mon = %d, mday = %d, wday = %d, hour = %d, "
+		"min = %d, sec = %d",
+		timeptr->tm_year, timeptr->tm_mon, timeptr->tm_mday, timeptr->tm_wday,
+		timeptr->tm_hour, timeptr->tm_min, timeptr->tm_sec);
+
+unlock:
+	k_mutex_unlock(&data->lock);
+
+	return err;
+}
+
+static const struct rtc_driver_api ds1307_driver_api = {
+	.set_time = ds1307_set_time,
+	.get_time = ds1307_get_time,
+};
+
+/*
+ * returns 1 : CH = 1; init failed
+ * returns 0 : CH = 0; init success
+ */
+static int ds1307_init(const struct device *dev)
+{
+	int err;
+	uint8_t ch;
+	uint8_t tx_buf[2];
+	const struct ds1307_config *config = dev->config;
+	struct ds1307_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+
+	if (!i2c_is_ready_dt(&config->i2c_bus)) {
+		LOG_ERR("I2C bus not ready");
+		return -ENODEV;
+	}
+
```

Comment:
Add an init of the control register here, ensuring no square wave is sent to any output. You can add configuration of the square wave to the devicetree if you want to, but without it, default to no square wave :)